.github/workflows/build.yml

```diff
@@ -25,60 +25,116 @@ jobs:
         run: |
-          . /home/firedrake/firedrake/bin/activate
-          firedrake-clean
-          export GUSTO_PARALLEL_LOG=FILE
-          export PYOP2_CFLAGS=-O0
-          python -m pip uninstall -y netCDF4
-          export HDF5_DIR=$PETSC_DIR/packages
-          export NETCDF4_DIR=$PETSC_DIR/packages
-          python -m pip install --no-binary netCDF4 --no-build-isolation netCDF4
-          python -m pytest \
-            -n 12 --dist worksteal \
-            --durations=100 \
-            --timeout=3600 \
-            --timeout-method=thread \
-            -o faulthandler_timeout=3660 \
-            -v unit-tests integration-tests examples
-        timeout-minutes: 120
+          . venv-gusto/bin/activate
+          : # Run the serial tests
+          firedrake-run-split-tests 1 1 -n 12 --verbose gusto-repo/integration-tests/model/test_nc_outputting.py
+          : # Run the parallel tests
+          firedrake-run-split-tests 2 6 --verbose gusto-repo/integration-tests/model/test_nc_outputting.py
+        timeout-minutes: 10
+
+      - name: Install parallel netCDF
+        # Run even if the step above failed
+        if: success() || steps.install-one.conclusion == 'success'
+        id: install-two
+        run: |
+          . venv-gusto/bin/activate
+          pip uninstall -y netCDF4
+          : # '--no-build-isolation' has to be passed for mpi4py to be found
+          : # but this means that we need to install any extra build
+          : # dependencies manually
+          pip install Cython
+          pip install --no-binary netCDF4 --no-build-isolation netCDF4
+
+      - name: Run tests (nprocs = 1)
+        run: |
+          . venv-gusto/bin/activate
+          : # Use pytest-xdist here so we can have a single collated output (not possible
+          : # for parallel tests)
+          firedrake-run-split-tests 1 1 -n 12 --dist worksteal "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 60
+
+      - name: Run tests (nprocs = 2)
+        # Run even if earlier tests failed
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 2 6 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 30
+
+      - name: Run tests (nprocs = 3)
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 3 4 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 10
+
+      - name: Run tests (nprocs = 4)
+        if: success() || steps.install-two.conclusion == 'success'
+        run: |
+          . venv-gusto/bin/activate
+          firedrake-run-split-tests 4 3 "$EXTRA_PYTEST_ARGS"
+        timeout-minutes: 10
+
       - name: Prepare logs
-        if: always()
+        if: success() || steps.install-two.conclusion == 'success'
+        id: prepare-logs
         run: |
           mkdir logs
           cd /tmp/pytest-of-firedrake/pytest-0/
```

Comment:
It looks like the location of the files has changed, maybe due to not being in a venv now? Where does pytest dump logs when running the docker container locally?

Reply:
But are they just the pytest logs? If so we have a better solution in Firedrake already. `firedrake-run-split-tests` outputs files called things like `pytest_nprocs1_job1.log` that we upload as artifacts [here](https://github.com/firedrakeproject/firedrake/blob/master/.github/workflows/build.yml#L182). Should we just do that here as well?

Reply:
I thought the gusto logs were also ending up there, but honestly can't remember off the top of my head. Something like if you call with xdist they end up there, maybe? I think it's best to check

Reply:
I've just looked at the logs from the latest run and the output looks quite confusing as each runner seems to be combining their logs without noting which test is being run. So yes you're right that the uploaded logs are not stdout but they're also very hard to understand.

This looks like it would be quite hard to recover this behaviour as we now call pytest about 20 times instead of just once so we have a lot more log files that can potentially conflict.

Reply:
I have a suggestion: if Gusto need to see the log information would it be acceptable to require them to set `GUSTO_PARALLEL_LOG=CONSOLE` in the `build.yml`? The output is then still available if needed.

I can leave a comment in `build.yml` to that effect.

Reply:
That definitely sounds acceptable to me. Does that mean that logged calls will not be uploaded as artifacts, but will be appear in the terminal-style output for the github action?

Reply:
Yes I think so. I will trigger a run now so we can see what that looks like.

To make things clear. The workflow I have *will* upload log artifacts at the end of a run. Those artifacts are just the output of *pytest*, not Gusto.

Reply:
Actually I think I may have a better solution. Give me a few minutes.

Reply:
Yep it works! Now there are 2 artifacts that get uploaded:
* pytest-logs: the pytest output from each job for each set of tests
* gusto-logs: the equivalent Gusto logging output